### PR TITLE
Issue #5018 - cancellation of WebSocketClient.connect Future should fail upgrade

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Issue #5018**

Cancelling the `Future<Session>` from `WebSocketClient.connect()` should cancel the websocket upgrade and close the websocket session if it has already been opened.

### CODE REVIEW

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (sbordet) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: Use just `1` second when you test for `false`.

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (sbordet) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: Strange comment :)
Why would `onOpen()` even be called, if the `connect()` has been canceled?

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (sbordet) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: Test also that the nested exception is a `CancellationException`?

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (sbordet) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: Remove `final` modifiers, unnecessary.

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (sbordet) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: This is probably going to fail in Jetty 10 since it's an anonymous endPoint and will fail the method handles permissions. Perhaps make it a proper `public` inner class now to avoid merge conflicts in the future?

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (sbordet) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: Please add a test case for the timeout of the future, i.e. calling `Future.get(1, SECONDS)`.
It should not fail the `connect()` since the semantic is just that of "1 second has passed" when it returns.
The test should then issue a second `Future.get()` that must succeed.

Also add a test that explicitly fails the future with a `CancellationException`, which should behave exactly like calling `Future.cancel()`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (lachlan-roberts) [reply jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: I didn't mean that `onOpen()` should be called, but that `onError()` should only be called if `onOpen()` has been called previously. I just wasn't expecting both of them to be notified of the failure, only the future. I will remove the comment.

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (lachlan-roberts) [reply jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: This shouldn't fail in Jetty 10, this `CloseTrackingEndpoint` extends the `WebSocketListener`. I think that issue with the public inner class is only for the annotated version as we need to discover the methods.

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (lachlan-roberts) [reply jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: I wrote the first test case.

For the second one there is no way to fail the future with a specific `Exception`. Unless we downcast to `CompletableFuture<Session>`.

Comment (sbordet) [reply jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: Please downcast.

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java
Comment (lachlan-roberts) [reply jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java]: Issue #5043 opened for this.

File: jetty-websocket/websocket-client/src/main/java/org/eclipse/jetty/websocket/client/WebSocketUpgradeRequest.java
Comment (sbordet) [inline jetty-websocket/websocket-client/src/main/java/org/eclipse/jetty/websocket/client/WebSocketUpgradeRequest.java]: Should not this be `if (throwable != null)`?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureTest.java
Comment (sbordet) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureTest.java]: Don't need this interface, use `Consumer<NativeWebSocketConfiguration>`

**Review by sbordet — APPROVED** [review]
Other than the nit in the test class, LGTM.

**Comment by lachlan-roberts** [comment]
Waiting on PR #5054 to be merged before merging this.

**Comment by joakime** [comment]
@lachlan-roberts PR #5054 has been merged.

**Pull request merged**
